README.md
binr
====

R package that provides algorithms for cutting numerical values
    exhibiting a potentially highly skewed distribution into evenly distributed
    groups (bins). This functionality can be applied for binning discrete
    values, such as counts, as well as for discretization of continuous values,
    for example, during generation of features used in machine learning
    algorithms.

See http://jabiru.github.io/binr/ for more details.


NEWS
Version 1.2
- Implementation of the predict() method by @smsaladi
- Handling of NAs by @smsalady
- Documentation updates
Version 1.1
- Fixed a bug in bins.getvals handling the case where number of bins is 1.
  Example:
      lst <- bins(c(rep(1,5), rep(0, 1000)), target.bins=3, minpts=10)
      bins.getvals(lst)

- Updated maintainer email and copyrights

Version 1.0
First public release.


R/greedy.R
#-------------------------------------------------------------------------------
#
# Package binr
#
# Implementation of greedy algorithm.
#
# Sergei Izrailev, 2011-2014
#-------------------------------------------------------------------------------
# Copyright 2011-2014 Collective, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------

#' \code{bins.greedy} - Wrapper around \code{bins.greedy.impl}. Goes over the
#' sorted values of \code{x} left to right and fills the bins with the values until
#' they are about the right size.
#' @param x Vector of numbers.
#' @param nbins Target number of bins.
#' @param minpts Minimum number of points in a bin. Only used if \code{naive = FALSE}.
#' @param naive When \code{TRUE}, simply calls \code{bins.greedy.impl} with data
#'        derived from \code{x}. Otherwise, makes an extra step of marking the values that
#'        by themselves take a whole bin to force the algorithm to place these values
#'        in a bin separately.
#' @name bins.greedy
#' @title Greedy binning algorithm.
#' @seealso \code{\link{binr}}, \code{\link{bins}}, \code{\link{bins.quantiles}} \code{\link{bins.optimize}}
#' @export bins.greedy
#' @rdname bins.greedy
#' @usage bins.greedy(x, nbins, minpts = floor(0.5 * length(x)/nbins), thresh = 0.8, naive = FALSE)
bins.greedy <- function(x, nbins, minpts = floor(0.5 * length(x)/nbins), thresh = 0.8, naive = FALSE)
{
   xtbl <- table(x)
   xval <- sort(unique(x))
   nvals <- length(xval)
   xstp <- vector(nvals, mode="logical")
   binsz <- floor(length(x) / nbins)
   if (naive) return( bins.greedy.impl(xval, xtbl, xstp, binsz, nbins, thresh) )

   ixdone <- which(xtbl >= binsz)
   ixkeep <- which(xtbl < binsz)
   nbins2 <- nbins - length(ixdone)
   nx2 <- sum(xtbl) - sum(xtbl[ixdone]) # how many points are left
   nbins.new <- min(nbins2, floor(nx2 / minpts) + 1) # make fewer bins if there's not enough points
   binsz.new <- floor(nx2 / nbins.new)
   istp <- ixdone + 1
   istp <- istp[istp < nvals]
   xstp.new <- xstp
   xstp.new[istp] <- TRUE
   #cuts.new <- bins.greedy.impl(xval[ixkeep], xtbl[ixkeep], xstp.new[ixkeep], binsz.new, nbins.new, thresh)
   cuts.new <- bins.greedy.impl(xval, xtbl, xstp, binsz.new, nbins.new, thresh)
   return(cuts.new)
}

#-------------------------------------------------------------------------------

#' \code{bins.greedy.impl} - Implementation of a single-pass binning algorithm that examines sorted data left to right
#' and builds bins of the target size. The \code{bins.greedy} wrapper around this function provides a less involved interface.
#' This is not symmetric wrt direction: symmetric distributions may not have symmetric bins if there are multiple points
#' with the same values. If a single value accounts for more than thresh * binsz points, it will be placed in
#' a new bin.
#' @param xval Sorted unique values of the data set x. This should be the numeric version of \code{names(xtbl)}.
#' @param xtbl Result of a call to \code{table(x)}.
#' @param xstp Stopping points; if \code{xstp[i] == TRUE}, the \code{i}-th value can't be merged to the \code{(i-1)}-th one.
#'        \code{xstp[1]} value is ignored.
#' @param binsz Target bin size, i.e., the number of points falling into each bin; for example, \code{floor(length(x) / nbins)}
#' @param thresh Threshold fraction of bin size for the greedy algorithm.
#'        Suppose there's \code{n < binsz} points in the current bin already.
#'        Also suppose that the next value V is represented by \code{m} points, and \code{m + n > binsz}.
#'        Then the algorithm will check if \code{m > thresh * binsz}, and if so, will place the value V into a new bin.
#'        If \code{m} is below the threshold, the points having value V are added to the current bin.
#' @param verbose When \code{TRUE}, prints the number of points falling into the bins.
#' @return A list with the following items:
#' \itemize{
#'    \item{binlo}{ - The "low" value falling into the bin.}
#'    \item{binhi}{ - The "high" value falling into the bin.}
#'    \item{binct}{ - The number of points falling into the bin.}
#'    \item{xtbl}{ - The result of a call to \code{table(x)}.}
#'    \item{xval}{ - The sorted unique values of the data points x. Essentially, a numeric version of \code{names(xtbl)}.}
#' }
#' @export bins.greedy.impl
#' @rdname bins.greedy
#' @usage bins.greedy.impl(xval, xtbl, xstp, binsz, nbins, thresh, verbose = F)
bins.greedy.impl <- function(xval, xtbl, xstp, binsz, nbins, thresh, verbose = F)
{
   nvals <- length(xval)
   if (nvals != length(xtbl) || nvals != length(xstp)) stop("bins.greedy: xval, xtbl and xstp lengths don't match")
   xstp[1] = FALSE
   binlo <- vector(nbins, mode="integer")
   binhi <- vector(nbins, mode="integer")
   binct <- vector(nbins, mode="integer")
   startbin <- TRUE
   k <- 1   # current bin index
   s <- 0   # running number of points in the current bin
   for (i in 1:nvals)
   {
      if (xstp[i]) # can't happen on i == 1
      {
         # wrap up the previous bin
         binhi[k] <- xval[i - 1]
         binct[k] <- s
         s <- 0
         k <- k + 1
         startbin <- TRUE
      }
      s <- s + xtbl[i]
      if (startbin)
      {
         binlo[k] <- xval[i]
         startbin <- FALSE
      }
      if (s >= binsz)
      {
         # we exceeded the bin size. If there were points in this bin before, check if
         # the added number of points exceeds the tolerance threshold.
         if (s > xtbl[i] && (xtbl[i] >= thresh * binsz))# || (s - xtbl[i] > thresh * binsz)))
         {
            # not the first item; complete the previous bin and make this a starting bin
            binhi[k] <- xval[i - 1]
            binct[k] <- s - xtbl[i]
            k <- k + 1
            s <- xtbl[i]
            binlo[k] <- xval[i]
            startbin <- FALSE
         }
         else
         {
            # done for now
            binhi[k] <- xval[i]
            binct[k] <- s
            k <- k + 1
            s <- 0
            startbin <- TRUE
         }
      }
   }
   if (!startbin)
   {
      # unfinished bin
      binhi[k] <- xval[nvals]
      binct[k] <- s
   }
   #if (k > nbins) k <- nbins  # we are NOT guaranteed that everything fits into at most nbins bins
   # Here's an example when things go bad:
   #    x <- rep(1:10, 10); bins.greedy(x, nbins=9, thresh=0.8)$binct
   # In this case, every value exceeds the threshold and is placed in a separate bin, hence we get
   # more bins than what we asked for.

   # Reduce k until a valid bin.
   while (is.na(binct[k])) k <- k - 1
   binlo <- binlo[1:k]
   binhi <- binhi[1:k]
   binct <- binct[1:k]

   names(binct) <- paste("[", binlo, ", ", binhi, "]", sep="")
   if (verbose) print(binct)
   return(list(binlo=binlo, binhi=binhi, binct=binct, xtbl=xtbl, xval=xval))
}

#-------------------------------------------------------------------------------


R/quantiles.R
#-------------------------------------------------------------------------------
#
# Package binr
#
# Implementation of an algorithm based on quantiles.
#
# Sergei Izrailev, 2011-2014
#-------------------------------------------------------------------------------
# Copyright 2011-2014 Collective, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


#' Cuts the data set x into roughly equal groups using quantiles.
#'
#' Because the number of unique values may be smaller than target.bins,
#' the function gradually increases the number of quantiles up to max.breaks
#' or until the target.bins number of bins is reached.
#' @param x A numeric vector to be cut in bins.
#' @param target.bins Target number of bins, which may not be reached if the
#'        number of unique values is smaller than the specified value.
#' @param max.breaks Maximum number of quantiles; must be at least as
#'        large as \code{target.bins}.
#' @param verbose Indicates verbose output.
#' @name bins.quantiles
#' @title Quantile-based binning
#' @seealso \code{\link{binr}}, \code{\link{bins}}, \code{\link{bins.greedy}}, \code{\link{bins.optimize}}
#' @export
bins.quantiles <- function(x, target.bins, max.breaks, verbose = FALSE)
{
   n.groups <- 1
   nbreak <- target.bins - 1
   if (max.breaks < target.bins) stop("bins.quantiles: max.breaks must be at least as large as target.bins")
   xtbl <- table(x)
   xval <- sort(unique(x))
   while (n.groups <= target.bins && nbreak <= max.breaks)
   {
      nbreak <- nbreak + 1
      qq <- unique(quantile(x, probs = 0:nbreak / nbreak))
      idx <- unique(findInterval(qq, xval))
      n.groups <- length(idx)
   }
   idx <- unique(findInterval(qq, xval))
   nbins <- length(idx) - 1
   binlo <- vector(nbins, mode="integer")
   binhi <- vector(nbins, mode="integer")
   binct <- vector(nbins, mode="integer")
   prev <- 0
   for (i in 1:nbins)
   {
      binlo[i] <- prev + 1
      binhi[i] <- idx[i + 1]
      binct[i] <- sum(xtbl[binlo[i]:binhi[i]])
      prev <- idx[i + 1]
   }
   names(binct) <- paste("[", xval[binlo], ", ", xval[binhi], "]", sep="")
   if (verbose) print(binct)
   return(list(binlo=binlo, binhi=binhi, binct=binct, xtbl=xtbl, xval=xval, err=sqrt(bins.merr(binct, target.bins))))
}

#-------------------------------------------------------------------------------


R/binr-package.R
#-------------------------------------------------------------------------------
#
# Package binr
#
# General description
#
# Sergei Izrailev, 2011-2014
#-------------------------------------------------------------------------------
# Copyright 2011-2014 Collective, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#-------------------------------------------------------------------------------
#' Package \code{binr} (pronounced as "binner") provides algorithms for cutting
#' numerical values exhibiting a potentially highly skewed distribution into
#' evenly distributed groups (bins). This functionality can be applied for
#' binning discrete values, such as counts, as well as for discretization of
#' continuous values, for example, during generation of features used in machine
#' learning algorithms.
#'
#' @name binr
#' @aliases binr
#' @seealso \code{\link{bins}}, \code{\link{bins.quantiles}}, \code{\link{bins.optimize}}, \code{\link{bins.greedy}}
#' @docType package
#' @title Cut Numeric Values Into Evenly Distributed Groups (bins).
#' @author Sergei Izrailev
#' @section Maintainer: Sergei Izrailev
#' @section Copyright: Copyright (C) Collective, Inc.; with portions Copyright (C) Jabiru Ventures LLC
#' @section License: Apache License, Version 2.0,
#'    available at http://www.apache.org/licenses/LICENSE-2.0
#' @section URL: http://github.com/jabiru/binr
#' @section Installation from github:
#' \code{devtools::install_github("jabiru/binr")}
#' @keywords csv delimited file read.csv bigint 64-bit integer64
#'
# The next and last line should be the word 'NULL'.
NULL


R/bins.R
#-------------------------------------------------------------------------------
#
# Package binr
#
# Implementation.
#
# Sergei Izrailev, 2011-2015
#-------------------------------------------------------------------------------
# Copyright 2011-2014 Collective, Inc.
# Copyright 2015 Jabiru Ventures LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#-------------------------------------------------------------------------------
#-------------------------------------------------------------------------------


#' Cut Numeric Values Into Evenly Distributed Groups (Bins)
#'
#' \code{bins} - Cuts points in vector x into evenly distributed groups (bins).
#' \code{bins} takes 3 separate approaches to generating the cuts, picks the one
#' resulting in the least mean square deviation from the ideal cut -
#' \code{length(x) / target.bins} points in each bin - and then merges small bins
#' unless \code{exact.groups} is \code{TRUE}
#' The 3 approaches are:
#' \enumerate{
#' \item{Use quantiles, and increase the number of even cuts up to max.breaks until the
#'    number of groups reaches the desired number. See \code{\link{bins.quantiles}}. }
#' \item{Start with a single bin with all the data in it and perform bin splits until
#'    either the desired number of bins is reached or there's no reduction in error
#'    (the latter is ignored if \code{exact.groups} is \code{TRUE}). See \code{\link{bins.split}}. }
#' \item{Start with \code{length(table(x))} bins, each containing exactly one distinct value
#'    and merge bins until the desired number of bins is reached. If \code{exact.groups} is
#'    \code{FALSE}, continue merging until there's no further reduction in error.
#'    See \code{\link{bins.merge}}. }
#' }
#' For each of these approaches, apply redistribution of points among existing bins
#' until there's no further decrease in error. See \code{bins.move}.
#'
#' The gains are computed using incremental analytical expresions derived for moving
#' a value from one bin to the next, splitting a bin into two or merging two bins.
#' @param x Vector of numbers
#' @param target.bins Number of groups desired; this is also the max number of groups.
#' @param max.breaks   Used for initial cut. If \code{exact.groups} is \code{FALSE}, bins are merged
#'                     until there's no bins with fewer than \code{length(x) / max.breaks} points.
#'                     In \code{bins}, one of \code{max.breaks} and \code{minpts} must be supplied.
#' @param exact.groups if TRUE, the result will have exactly the number of target.bins;
#'                     if FALSE, the result may contain fewer than target.bins bins
#' @param verbose      Indicates verbose output.
#' @param errthresh    If the error is below the provided value, stops after the first rough estimate of the bins.
#' @param minpts       Minimum number of points in a bin.
#'                     In \code{bins}, one of \code{max.breaks} and \code{minpts} must be supplied.
#' @return A list containing the following items (not all of them may be present):
#' \itemize{
#'    \item{binlo}{ - The index into \code{xval} yielding the "low" value falling into the bin.}
#'    \item{binhi}{ - The index into \code{xval} yielding the "high" value falling into the bin.}
#'    \item{binct}{ - The number of points falling into the bin.}
#'    \item{xtbl}{ - The result of a call to \code{table(x)}.}
#'    \item{xval}{ - The sorted unique values of the data points x. Essentially, a numeric version of \code{names(xtbl)}.}
#'    \item{changed}{ - Flag indicating whether the bins have been modified by the function.}
#'    \item{err}{ - Mean square root error between the resulting counts and ideal bins.}
#'    \item{imax}{ - For the move, merge and split operations, the index of the bin with the maximum gain.}
#'    \item{iside}{ - For the move operation, the side of the move: 0 = left, 1 = right.}
#'    \item{gain}{ - Error gain obtained as the result of the function call.}
#' }
#'
#' @examples
#' \dontrun{
#'    # Seriously skewed x:
#'    x <- floor(exp(rnorm(200000 * 1.3)))
#'    cuts <- bins(x, target.bins = 10, minpts = 2000)
#'    cuts$breaks <- bins.getvals(cuts)
#'    cuts$binct
#'    #   [0, 0]    [1, 1]    [2, 2]    [3, 3]    [4, 4]    [5, 5]    [6, 7]   [8, 10]
#'    # 129868     66611     28039     13757      7595      4550      4623      2791
#'    #   [11, 199]
#'    # 2166
#'
#'    # Centered x:
#'    x <- rep(c(1:10,20,31:40), c(rep(1, 10), 100, rep(1,10)))
#'    cuts <- bins(x, target.bins = 3, minpts = 10)
#'    cuts$binct
#'    # [1, 10] [20, 20] [31, 40]
#'    #      10      100       10
#' }
#' @seealso \code{\link{binr}}, \code{\link{bins.greedy}}, \code{\link{bins.quantiles}} \code{\link{bins.optimize}}
#' @export
#' @rdname bins
bins <- function(x, ...) {
   UseMethod("bins")
}

#' @export
#' @rdname bins
bins.default <- function(x, target.bins, max.breaks = NA, exact.groups=F, verbose=F, errthresh = 0.1, minpts = NA)
{
   x <- x[!is.na(x)]
   if (length(x) < target.bins) stop(paste("bins: number of desired groups (", target.bins, ") is greater than the number of points (", length(x), ")"))
   all <- vector(3, mode="list")
   names(all) <- c("quantile", "split", "merge")
   if (is.na(minpts))
   {
      if (is.na(max.breaks) || max.breaks < 1)
      {
         stop(paste("bins: missing or invalid max.breaks = ", max.breaks, " and minpts = ", minpts))
      }
      minpts <- floor(length(x) / max.breaks)
   } else if (is.na(max.breaks))
   {
      if (minpts < 1) stop(paste("bins: invalid minpts = ", minpts))
      max.breaks <- ceiling(length(x) / minpts)
   } else
   {
      stop("bins: only one of max.breaks and minpts can be specified")
   }

   all$quantile <- bins.quantiles(x, target.bins, max.breaks)
   if (all$quantile$err / (length(x) / target.bins) < errthresh)
      return(structure(all$quantile, class = "binr")) # good enough

   xval <- all$quantile$xval
   xtbl <- all$quantile$xtbl

   all$split <- list(binlo = 1, binhi = length(xval), binct = sum(xtbl), xval = xval, xtbl = xtbl)
   #all$merge <- list(binlo = 1:length(xtbl), binhi = 1:length(xtbl), binct = xtbl, xval = xval, xtbl = xtbl)
   all$merge <- bins.quantiles(x, max.breaks * 2, max.breaks * 2) # start with a fairly fine breakup

   all$quantile <- bins.move.iter(all$quantile, target.bins, verbose=verbose)
   if (exact.groups)
   {
      if (length(all$quantile$binct) < exact.groups)
      {
         all$quantile <- bins.split.iter(all$quantile, target.bins, verbose=verbose, exact.groups=T)
         all$quantile <- bins.move.iter(all$quantile, target.bins, verbose=verbose)
      }
      else if (length(all$quantile$binct) > exact.groups)
      {
         all$quantile <- bins.merge.iter(all$quantile, target.bins, verbose=verbose, exact.groups=T)
         all$quantile <- bins.move.iter(all$quantile, target.bins, verbose=verbose)
      }
   }

   # don't attempt to split or merge if the number of values is too big
   if (length(xval) < 10000)
   {
      # split
      all$split <- bins.split.iter(all$split, target.bins, verbose=verbose, exact.groups=exact.groups)
      all$merge <- bins.merge.iter(all$merge, target.bins, verbose=verbose, exact.groups=exact.groups)

      all <- lapply(all, function(lst) bins.move.iter(lst, target.bins, verbose=verbose))
      errs <- unlist(lapply(all, function(lst) sqrt(bins.merr(lst$binct, target.bins))))
      names(errs) <- names(all)
      if (verbose) print(errs)
      imin <- which(errs == min(errs))[1]

      lst <- all[[imin]]
   }
   else
   {
      lst <- bins.move.iter(all$quantile, target.bins, verbose=verbose)
   }

   if (!exact.groups)
   {
      # merge small bins
      while (any(lst$binct < minpts))
      {
         lst <- bins.merge(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct, target.bins=target.bins, force=T, verbose=verbose)
         lst <- bins.move.iter(lst, target.bins, verbose=verbose)
         if (verbose) print(paste(lst$gain, lst$err))
      }
   }

   structure(lst, class = "binr")
}

#' @export
#' @rdname bins
bins.data.frame <- function(df, ...) {
   structure(
     lapply(df, function(x, ...) bins(x, ...), ...),
     class = "binr"
     )
}

#' @export
#' @rdname bins
predict.binr <- function(obj, data, labels = FALSE, ...) {
   if (is.data.frame(data)) {
     if (all(names(data) %in% names(obj))) {
       return(as.data.frame(sapply(
         names(data), function(x) predict(obj[[x]], data[[x]],
         labels = labels, ...)
         )))
     } else {
        stop("data.frame has columns without bin cuts")
     }
   } else {
     out <- cut(data, bins.getvals(obj),
                labels = if(labels) NULL else FALSE, ...)
     if (NA %in% out) {
       if (labels) {
         levels(out) <- c(levels(out), "[NA]")
         out[is.na(out)] <- "[NA]"
       } else {
         out[is.na(out)] <- max(out, na.rm = TRUE) + 1
       }
     }
     return(out)
   }
}

#-------------------------------------------------------------------------------

#' \code{bins.getvals} - Extracts cut points from the object retured by \code{bins}.
#' The cut points are always between the values in \code{x} and weighed such that
#' the cut point splits the area under the line from (lo, n1) to (hi, n2) in half.
#' @param lst The list returned by the \code{bins} function.
#' @param minpt The value replacing the lower bound of the cut points.
#' @param maxpt The value replacing the upper bound of the cut points.
#' @return \code{bins.getvals} returns a vector of cut points extracted from the
#'         \code{lst} object. The actual low and high values in each bin, as well
#'         as the counts of values in each bin are placed in attributes
#'         \code{binlo}, \code{binhi} and \code{binct}, respectively.
#' @export bins.getvals
#' @rdname bins
#' @usage bins.getvals(lst, minpt = -Inf, maxpt = Inf)
bins.getvals <- function(lst, minpt = -Inf, maxpt = Inf)
{
   # finds the point that splits the area under the line from (lo, n1) to (hi, n2) in half.
   cutpt <- function(lo, hi, n1, n2)
   {
      if (n1 == n2) return((hi + lo) / 2)
      a <- hi - lo
      b <- 2 * a * n1 / (n2 - n1)
      c = a * a * (n1 + n2)  / (n2 - n1) / 2
      d <- (-b + sign(n2 - n1) * sqrt(b * b + 4 * c)) / 2
      return(lo + d)
   }

   nbins <- length(lst$binct)
   if (nbins == 0) stop("bins.getvals: zero bins")
   res <- vector(nbins + 1, mode="double")
   res[1] <- minpt
   res[nbins + 1] <- maxpt

   if (nbins > 1)
   {
      for (i in 2:nbins)
      {
         res[i] <- cutpt(lst$xval[lst$binhi[i - 1]], lst$xval[lst$binlo[i]], lst$binct[i - 1], lst$binct[i])
      }
      names(res) <- c(paste("[", res[1:(nbins-1)], ", ", res[2:nbins], ")", sep=""), paste("[", res[nbins], ", ", res[nbins + 1], "]", sep=""), "LAST")
   } else {
      # nbins = 1
      names(res) <- c(paste("[", res[1], ", ", res[2], "]", sep=""), "LAST")
   }

   attr(res, "binlo") <- lst$xval[lst$binlo]
   attr(res, "binhi") <- lst$xval[lst$binhi]
   attr(res, "binct") <- lst$binct
   return(res)
}


#-------------------------------------------------------------------------------

#' \code{bins.merr} - Partitioning the data into bins using splitting, merging
#' and moving optimizes this error function, which is the mean squared error
#' of point counts in the bins relative to the optimal number of points per bin.
#' @param binct The number of points falling into the bins.
#' @export bins.merr
#' @rdname bins
#' @usage bins.merr(binct, target.bins)
bins.merr <- function(binct, target.bins) {
  mean((binct - sum(binct) / target.bins)^2)
}

#-------------------------------------------------------------------------------

if (0)
{
   xnorm <- rnorm(200000)
   xlog <- floor(exp(xnorm * 1.3))
   xlog2 <- c(xlog, -xlog - 1)
   x <- xlog2

   xtbl <- table(x)
   xval <- sort(unique(x))
   nvals <- length(xval)
   nbins <- 10

   binsz <- floor(length(x) / nbins)
   minpts <- 1000
   thresh <- 0.8
   target.bins = 10
   max.breaks = 20

   binlo <- 1:10
   binhi <- c(1:9,length(xval))
   binct <- c(xtbl[1:9],sum(xtbl[10:length(xtbl)]))
   source("~/cpp/cmrutils/R/cmcut.R")
   lst <- adjust.cells(xval, xtbl, binlo, binhi, binct)
   bc <- rep(0,length(lst$binct)); while(!all(lst$binct == bc)) { bc <- lst$binct; lst <- adjust.cells(xval, xtbl, lst$binlo, lst$binhi, lst$binct) }
   lst <- split.cells(xval, xtbl, lst$binlo, lst$binhi, lst$binct); lst
   prevlen <- 0; while(length(lst$binct) != prevlen) { prevlen <- length(lst$binct); lst <- split.cells(xval, xtbl, lst$binlo, lst$binhi, lst$binct) }

   lst <- bins.quantiles(x, 10, 20)
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.move(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }

   lst$binlo <- 1:10
   lst$binhi <- c(1:9,length(xval))
   lst$binct <- c(xtbl[1:9],sum(xtbl[10:length(xtbl)]))

   lst$binlo <- 1; lst$binhi <- length(xval); lst$binct <- sum(xtbl)
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.split(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct, target.bins); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.move(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.merge(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct, target.bins); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.move(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }


   lst$binlo <- 1; lst$binhi <- length(xval); lst$binct <- sum(xtbl)
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.split(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct, 10); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }
   while(length(lst$binct) < 10) { lst <- bins.split(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct, target.bins, force=T) }
   preverr <- Inf; err <- sqrt(bins.merr(lst$binct, target.bins)); while (preverr > err) { preverr <- err; lst <- bins.split(lst$xval, lst$xtbl, lst$binlo, lst$binhi, lst$binct, 10); err <- sqrt(bins.merr(lst$binct, target.bins)); print(paste(lst$gain, err)); }
 }


#-------------------------------------------------------------------------------
